FlashLFQ/Bayesian Protein Quantification/ProteinQuantificationEngine.cs

```diff
@@ -0,0 +1,604 @@
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace FlashLFQ
+{
+    /// <summary>
+    /// This is the Bayesian protein quantification engine used by FlashLFQ. It uses a Markov Chain Monte Carlo (MCMC)
+    /// algorithm to estimate the fold-change of a protein between conditions, given its constituent peptides fold-change measurements.
+    /// This estimation is performed by fitting Student's t-distribution to the peptide fold-change data.
+    /// 
+    /// The posterior error probability (PEP) of a protein is the probability that the protein's fold-change is below a certain value ("cutoff").
+    /// This "cutoff" can be defined by the user or can be estimated from the data using an "experimental null" strategy.
+    /// The PEP is estimated using a "skeptical prior probability", which means that the initial assumption (given no data) is that the 
+    /// protein is not changing in abundance between conditions. The number of data points (peptide fold-changes for a protein), the 
+    /// magnitude of the fold-change relative to the cutoff, and the precision of the data points all factor in to this PEP calculation.
+    /// 
+    /// In other words, the default assumption is that the protein is not changing; it must be "proven" to be changing by examining the evidence.
+    /// 
+    /// See the BayesianEstimation project for details of the MCMC algorithm.
+    /// </summary>
+    public class ProteinQuantificationEngine
+    {
+        private readonly FlashLfqResults results;
+        private readonly int MaxThreads;
+        private readonly bool UseSharedPeptides;
+        private readonly bool PairedSamples;
+        private readonly string BaseCondition;
+        private readonly double? FoldChangeCutoff;
+        private readonly int BurnInSteps;
+        private readonly int NSteps;
+        private readonly MersenneTwister Rng;
+        private bool conditionsAreDefined;
+        private Dictionary<ProteinGroup, List<Peptide>> ProteinsWithConstituentPeptides;
+        private Dictionary<(Peptide, string, int), double> PeptideToSampleQuantity;
+        private Dictionary<ProteinGroup, double> ProteinToBaseConditionIntensity;
+
+        /// <summary>
+        /// Constructs the protein quantification engine. 
+        /// 
+        /// "baseCondition" refers to the condition that the protein fold-changes could be calculated relative to.
+        /// 
+        /// If "foldChangeCutoff" is null, the fold-change cutoff will be estimated from the data ("experimental null").
+        /// 
+        /// </summary>
+        public ProteinQuantificationEngine(FlashLfqResults results, int maxThreads, string baseCondition, bool useSharedPeptides = false,
+            double? foldChangeCutoff = null, int? randomSeed = null, int mcmcBurninSteps = 1000, int mcmcSteps = 3000, bool pairedSamples = false)
+        {
+            if (string.IsNullOrWhiteSpace(baseCondition))
+            {
```

Comment:
Similar to the last comment. Is a baseCondition necessary? Couldn't you arbitrarily grab one of the conditions, treat it as the "base"/WT, and get similar results?

Reply:
no. the fold-changes are relative to something. if you choose a different reference, you'll get a different answer

Reply:
Can you clarify really quick: If there are two conditions, A and B, do you get different results when A is the base vs when B is the base, even though you're comparing the same conditions?

Reply:
sorry, yeah, that was unclear. if you do A vs. B and then B vs. A, you'll get the same answer (approximately, given that there is inherent randomness in the MCMC sampler). if you do:

A vs. B
C vs. B

you'll get a different estimate for the fold-change (and intensity) of B, because the B condition has different references.

Reply:
I see. If you wanted to do an all-all comparison, then you would end up doing (A vs B, A vs C) and (B vs A, B vs C), which works but is computationally redundant.

Reply:
yeah. I suppose eventually I could implement a Bayesian ANOVA. that's outside the scope of this PR though